src/profile-logic/process-profile.js

```diff
@@ -1675,3 +1686,170 @@ export async function unserializeProfileOfArbitraryFormat(
+ * to the parent process and tab process main threads to show the visual progress.
+ * Mutates the markers inside parent process and tab process main threads.
+ */
+export function processVisualMetrics(
+  threads: Thread[],
+  meta: ProfileMeta,
+  pages: PageList
+) {
+  const { visualMetrics } = meta;
+  if (pages.length === 0 || !visualMetrics) {
+    // No pages or visualMetrics were found in the profile. Skip this step.
+    return;
+  }
+
+  // Find the parent process and the tab process main threads.
+  const mainThreadIdx = threads.findIndex(
+    (thread) => thread.name === 'GeckoMain' && thread.processType === 'default'
+  );
+  const tabThreadIdx = findTabMainThreadForVisualMetrics(threads, pages);
+
+  if (mainThreadIdx === -1 || !tabThreadIdx) {
+    // Failed to find the parent process or tab process main threads. Return early.
+    return;
+  }
+  const mainThread = threads[mainThreadIdx];
+  const tabThread = threads[tabThreadIdx];
+
+  // These metrics are currently present inside profile.meta.visualMetrics.
+  const metrics = ['Visual', 'ContentfulSpeedIndex', 'PerceptualSpeedIndex'];
+  // Find the Test category so we can add the visual metrics markers with it.
+  if (meta.categories === undefined) {
+    // Making Flow happy. This means that this is a very old profile.
+    return;
+  }
+  const testingCategoryIdx = meta.categories.findIndex(
+    (cat) => cat.name === 'Test'
+  );
+
+  function addMetricMarker(
+    thread: Thread,
+    name: string,
+    startTime: number,
+    endTime?: number,
+    payload?: any
+  ) {
+    // Add the marker to the given thread.
+    thread.markers.name.push(thread.stringTable.indexForString(name));
+    thread.markers.startTime.push(startTime);
+    thread.markers.endTime.push(endTime ? endTime : 0);
+    thread.markers.phase.push(endTime ? INTERVAL : INSTANT);
+    thread.markers.category.push(testingCategoryIdx);
+    thread.markers.data.push(payload ?? null);
+    thread.markers.length++;
+  }
+
+  for (const metricName of metrics) {
+    const metric = visualMetrics[`${metricName}Progress`];
+    if (!metric) {
+      // Skip it if we don't have this metric.
+      continue;
+    }
+
+    const startTime = metric[0].timestamp;
+    const endTime = metric[metric.length - 1].timestamp;
+
+    // Add the progress marker to the parent process main thread.
+    const markerName = `${metricName} Progress`;
+    addMetricMarker(mainThread, markerName, startTime, endTime);
+    // Add the progress marker to the tab process main thread.
+    addMetricMarker(tabThread, markerName, startTime, endTime);
+
+    // Add progress markers for every visual progress change for more fine grained information.
+    const progressMarkerSchema = {
+      name: 'VisualMetricProgress',
+      tableLabel: '{marker.name} — {marker.data.percentage}',
+      display: ['marker-chart', 'marker-table', 'timeline-overview'],
+      data: [{ key: 'percentage', label: 'Percentage', format: 'percentage' }],
```

Comment:
The thing is that, I think we need to add a different description entry for each metric type so we can explain what a specific metric mean. I tried to come up with small and and understandable strings but I couldn't. It could be good to get help from people who are more familiar with them. Maybe from the perftests team?

Reply:
Maybe browsertime folks have good explanations? It's good to do that as a follow-up anyway.